src/Main.elm

```diff
@@ -1,71 +1,427 @@
+        |> request
+
+
+postTodo : String -> Cmd Msg
+postTodo title =
+    post "https://todo-backend-hanami.herokuapp.com/"
+        |> withBody
+            (Http.jsonBody
+                (Encode.object
+                    [ ( "title", Encode.string title ) ]
+                )
+            )
+        |> withExpect (Http.expectJson AddTodoCmdComplete todoDecoder)
+        |> request
+
+
+updateTodo : Todo -> Bool -> String -> Cmd Msg
+updateTodo todo isCompleted title =
+    patch todo.url
+        |> withBody
+            (Http.jsonBody
+                (Encode.object
+                    [ ( "completed", Encode.bool isCompleted )
+                    , ( "title", Encode.string title )
+                    ]
+                )
+            )
+        |> withExpect (Http.expectJson UpdateCmdComplete todoDecoder)
+        |> request
+
+
+deleteTodo : Todo -> Cmd Msg
+deleteTodo todo =
+    delete todo.url
+        |> withExpect (Http.expectWhatever DeleteTodoCmdComplete)
+        |> request
+
+
+focusRenameInput : Cmd Msg
+focusRenameInput =
+    Task.attempt (always NoOp) (Dom.focus "todo-rename-input")
+
+
+addTodo : Model -> ( Model, Cmd Msg )
+addTodo model =
+    ( { model | newTodoTitle = "" }, postTodo model.newTodoTitle )
+
+
+type Msg
+    = AddTodo
+    | AddTodoCmdComplete (Result Http.Error Todo)
+    | DeleteTodo Todo
+    | DeleteTodoCmdComplete (Result Http.Error ())
+    | EndRenaming Todo
+    | FetchTodosCmdComplete (Result Http.Error (List Todo))
+    | HideCompleted Bool
+    | InputNewTodoTitle String
+    | KeyDown KeyDownWhere Int
+    | NoOp
+    | RandomIndexComplete Int
+    | RenameTodo Todo String
+    | StartRenaming Todo
+    | UpdateChecked Todo Bool
+    | UpdateCmdComplete (Result Http.Error Todo)
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
 update msg model =
     case msg of
+        AddTodo ->
+            addTodo model
+
+        AddTodoCmdComplete (Ok _) ->
+            ( model, fetchTodos )
+
+        DeleteTodo theTodo ->
+            ( { model
+                | todos =
+                    List.filter
+                        (\todo ->
+                            todo.id /= theTodo.id
+                        )
+                        model.todos
+              }
+            , deleteTodo theTodo
+            )
+
+        DeleteTodoCmdComplete (Err _) ->
+            ( model, fetchTodos )
+
+        EndRenaming theTodo ->
+            ( { model
+                | todoBeingRenamed = Nothing
+              }
+            , updateTodo theTodo theTodo.completed theTodo.title
+            )
+
+        FetchTodosCmdComplete (Ok todos) ->
+            ( { model | todos = todos, state = Normal }, Cmd.none )
+
+        HideCompleted checkedUnchecked ->
+            ( { model | hideCompleted = checkedUnchecked }, Cmd.none )
+
+        InputNewTodoTitle title ->
+            ( { model | newTodoTitle = title }, Cmd.none )
+
+        KeyDown OnAdd 13 ->
```

Comment:
Would have been great to have some constants for this keys, 13 is like a magic number code smell.

Reply:
Haha, right. Lucky number 13. Will see how I can make obvious what it means (the Enter key).